File: server/src/services/skillInference.js
const SKILL_ALIASES = [
  {
    pattern: /\bmern(?:\s+stack)?(?:\s+developer|\s+engineer)?\b/i,
    skills: ['mongodb', 'express', 'react', 'node.js', 'javascript']
  },
  {
    pattern: /\bmean(?:\s+stack)?(?:\s+developer|\s+engineer)?\b/i,
    skills: ['mongodb', 'express', 'angular', 'node.js', 'javascript']
  },
  {
    pattern: /\bmevn(?:\s+stack)?(?:\s+developer|\s+engineer)?\b/i,
    skills: ['mongodb', 'express', 'vue', 'node.js', 'javascript']
  },
  {
    pattern: /\bml(?:\s+backend)?(?:\s+engineer|\s+developer)?\b|\bmachine learning backend(?:\s+engineer|\s+developer)?\b/i,
    skills: ['python', 'machine learning', 'api', 'fastapi', 'sql']
  },
  {
    pattern: /\bdata scientist\b/i,
    skills: ['python', 'sql', 'machine learning', 'statistics', 'data analysis']
  },
  {
    pattern: /\bdevops(?:\s+engineer)?\b/i,
    skills: ['docker', 'kubernetes', 'ci/cd', 'cloud']
  }
];

const normalizeSkill = value => String(value || '').trim().toLowerCase();

const unique = values => [...new Map(
  values
    .map(value => String(value || '').trim())
    .filter(Boolean)
    .map(value => [normalizeSkill(value), value])
).values()];

const inferSkillsFromText = text => {
  const value = String(text || '');
  const inferred = [];

  for (const alias of SKILL_ALIASES) {
    if (alias.pattern.test(value)) {
      inferred.push(...alias.skills);
    }
  }

  return unique(inferred);
};

const expandSkills = (skills = [], contextText = '') => {
  const explicitSkills = Array.isArray(skills)
    ? skills.map(skill => String(skill || '').trim()).filter(Boolean)
    : [];
  const inferredFromSkillNames = explicitSkills.flatMap(inferSkillsFromText);
  const inferredFromContext = inferSkillsFromText(contextText);

  return unique([...explicitSkills, ...inferredFromSkillNames, ...inferredFromContext]);
};

const enrichTextWithInferredSkills = text => {
  const inferred = inferSkillsFromText(text);
  if (!inferred.length) return String(text || '');
  return `${text || ''}\nInferred skills: ${inferred.join(', ')}`;
};

module.exports = {
  expandSkills,
  inferSkillsFromText,
  enrichTextWithInferredSkills
};

File: hiring-agent/backend/utils/skill_inference.py
from __future__ import annotations

import re


SKILL_ALIASES: list[tuple[re.Pattern[str], list[str]]] = [
    (
        re.compile(r"\bmern(?:\s+stack)?(?:\s+developer|\s+engineer)?\b", re.I),
        ["mongodb", "express", "react", "node.js", "javascript"],
    ),
    (
        re.compile(r"\bmean(?:\s+stack)?(?:\s+developer|\s+engineer)?\b", re.I),
        ["mongodb", "express", "angular", "node.js", "javascript"],
    ),
    (
        re.compile(r"\bmevn(?:\s+stack)?(?:\s+developer|\s+engineer)?\b", re.I),
        ["mongodb", "express", "vue", "node.js", "javascript"],
    ),
    (
        re.compile(
            r"\bml(?:\s+backend)?(?:\s+engineer|\s+developer)?\b"
            r"|\bmachine learning backend(?:\s+engineer|\s+developer)?\b",
            re.I,
        ),
        ["python", "machine learning", "api", "fastapi", "sql"],
    ),
    (
        re.compile(r"\bdata scientist\b", re.I),
        ["python", "sql", "machine learning", "statistics", "data analysis"],
    ),
    (
        re.compile(r"\bdevops(?:\s+engineer)?\b", re.I),
        ["docker", "kubernetes", "ci/cd", "cloud"],
    ),
]


def _unique(values: list[str]) -> list[str]:
    seen: set[str] = set()
    unique_values: list[str] = []

    for value in values:
        cleaned = value.strip()
        key = cleaned.lower()
        if cleaned and key not in seen:
            seen.add(key)
            unique_values.append(cleaned)

    return unique_values


def infer_skills_from_text(text: str) -> list[str]:
    inferred: list[str] = []
    for pattern, skills in SKILL_ALIASES:
        if pattern.search(text or ""):
            inferred.extend(skills)
    return _unique(inferred)


def expand_skills(skills: list[str], context_text: str = "") -> list[str]:
    explicit = [skill.strip() for skill in skills if skill.strip()]
    inferred_from_skills: list[str] = []
    for skill in explicit:
        inferred_from_skills.extend(infer_skills_from_text(skill))

    return _unique([
        *explicit,
        *inferred_from_skills,
        *infer_skills_from_text(context_text),
    ])
